Bugsnag/Assets/Bugsnag/Editor/SymbolUpload/BugsnagAndroidSymbolUploader.cs

```diff
@@ -0,0 +1,246 @@
+        {
+
+            // Only upload symbols for Android builds
+            if (report.summary.platform != BuildTarget.Android)
+            {
+                return;
+            }
+
+            // Get the Bugsnag settings object and check if auto-upload is enabled
+            var config = GetSettingsObject();
+            if (config == null || !config.AutoUploadSymbols)
+            {
+                return;
+            }
+
+            UnityEngine.Debug.Log("Beginning Bugsnag Android automatic symbol upload.");
+            EditorUtility.DisplayProgressBar("Bugsnag Symbol Upload", "Uploading Android Symbol Files", 0.0f);
+
+            // Check if symbol creation is enabled
+            if (EditorUserBuildSettings.androidCreateSymbols != AndroidCreateSymbols.Public && EditorUserBuildSettings.androidCreateSymbols != AndroidCreateSymbols.Debugging)
+            {
+                UnityEngine.Debug.LogError("Bugsnag symbol upload is enabled but Android symbol creation is disabled. Please enable symbol creation in your build settings.");
+                EditorUtility.ClearProgressBar();
+                return;
+            }
+
+            // check if the API key is set
+            var apiKey = config.ApiKey;
+            if (string.IsNullOrEmpty(apiKey))
+            {
+                UnityEngine.Debug.LogError("Bugsnag symbol upload is enabled but the API key is not set. Please set the API key in the Bugsnag settings window.");
+                EditorUtility.ClearProgressBar();
+                return;
+            }
+
+            // Check if a cli path was provided in the settings object, if yes use it if not download the cli
+            string cliExecutablePath;
+            if (!string.IsNullOrEmpty(config.BugsnagCLIExecutablePath))
+            {
+                cliExecutablePath = config.BugsnagCLIExecutablePath;
+            }
+            else
+            {
+                if (!DownloadBugsnagCli(GetBugsnagCliDownloadUrl(), _cliDownloadPath) || !MakeExecutable(_cliDownloadPath))
+                {
+                    UnityEngine.Debug.LogError("Failed to download and make the Bugsnag CLI executable.");
+                    EditorUtility.ClearProgressBar();
+                    return; // Exit if the download or chmod process failed
+                }
+                cliExecutablePath = _cliDownloadPath;
+            }
+
+
+            var buildOutputPath = Path.GetDirectoryName(report.summary.outputPath);
+
+            var args = string.Format("upload unity-android --api-key={0} --verbose --project-root={1} {2}", apiKey, Application.dataPath, buildOutputPath);
+            if (!string.IsNullOrEmpty(config.AppVersion))
+            {
+                args += $" --version-name={config.AppVersion}";
+            }
+            if (config.VersionCode > -1)
+            {
+                args += $" --version-code={config.VersionCode}";
+            }
+            RunBugsnagCliCommand(cliExecutablePath, args);
```

Comment:
I realise this is all just internal, but I'd suggest it would be better to push all the CLI arg handling into the execute function: just passing the outputPath from the report to the function.

Then a) that function alone deals with how to call the CLI and how to use configuration to apply to arguments; and b) when it comes to the iOS part, you should be able to reuse a lot of this functionality and the platform condition becomes whether to call `RunBugsnagCliCommand` (Android) or a new `UpdateXcodeProjectToRunCLI` (iOS).

Reply:
Agreed, i have refactored it